azurelinuxagent/common/protocol/wire.py

```diff
@@ -768,6 +796,35 @@ def _update_from_goal_state(self, refresh_type):
 
         raise ProtocolError("Exceeded max retry updating goal state")
 
+    def _save_goal_state_if_necessary(self, new_goal_state, refresh_type):
+        # We cache the goal state if one of the following is true
+        # 1) We didn't previously have a goal state
+        # 2) The incarnation changed
+        # 3) The extensions config changed
+        # 4) The extensions config did not change, but previously it did
+        # 5) We have an update type of GoalStateForced
+        save_goal_state = False
+        if new_goal_state is None:
+            return
+        if self._goal_state is None or new_goal_state.changed:
```

Comment:
Apart from this place, I couldn't find any other occurance of `GoalState.changed` property. 
From what I gathered, it just checks if incarnation changed which would also be checked by `ext_conf.changed` property. If that assumption is correct, could you remove the `GoalState.changed` property as its not being used anywhere else.

Reply:
The problem is we're not guaranteed to have an extensionsConfig, so we still need to have an incarnation. This is explained in the comment above this and where we set goal_state.changed.

Reply:
Sorry missed that. I have a follow up point then, wouldn't it make more sense to have something like this instead - 
`if self._goal_state is None or new_goal_state.incarnation != self._goal_state.incarnation`. Basically I want to avoid adding another property if its not being used anywhere else, minimising the extra stuff as much as possible :)

Reply:
That may be possible, but I wanted to avoid checking for incarnation outside of goal_state or goal_state_retriever. That's a detail internal to the class and I would like to avoid using it outside of these two classes because regressions are then likely. The code outside of goal_state should just care whether it changed.

Reply:
I respectfully disagree :)
`changed` is not a property of the GoalState, its a state change that we maintain to see if the incarnation changed from our previously cached state. On the other hand, `incarnation` is a property of the GoalState file (that is generated from CRP) which is something we should parse and keep accordingly. 

I totally agree with this statement - 
>but I wanted to avoid checking for incarnation outside of goal_state or goal_state_retriever

In this case something the interface Norberto added makes more sense- Eg: `GoalState.fetch_full_goal_state_if_incarnation_different_than` which keeps the incarnation state internal (kinda) and reduces the scope for errors. 

I wont ask you to make any more additional breaking changes that would cause a lot of pain, but I would appreciate it if you removed the `changed` property. Thanks :)

Reply:
Property removed